app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,598 @@
+  public static int FILE_TRANSFER_PORT = 8008;
+
+  @Inject SharedPreferenceUtil sharedPreferenceUtil;
+  @Inject AlertDialogShower alertDialogShower;
+
+  @BindView(R.id.toolbar_local_file_transfer) Toolbar actionBar;
+  @BindView(R.id.text_view_device_name) TextView deviceName;
+  @BindView(R.id.progress_bar_searching_peers) ProgressBar searchingPeersProgressBar;
+  @BindView(R.id.list_peer_devices) ListView listViewPeerDevices;
+  @BindView(R.id.text_view_empty_peer_list) TextView textViewPeerDevices;
+  @BindView(R.id.recycler_view_transfer_files) RecyclerView filesRecyclerView;
+
+  private ArrayList<Uri> fileUriArrayList; // For sender device, stores uris of the files
+  public @NonNull Boolean fileSendingDevice = false;// Whether the device is the file sender or not
+
+  public @NonNull WifiDirectManager wifiDirectManager = new WifiDirectManager(this);
+
+  private int totalFilesForTransfer = -1;
+  private int filesSent = 0;          // Count of number of files transferred until now
+  private ArrayList<FileItem> filesToSend = new ArrayList<>();
+  private FileListAdapter fileListAdapter;
+
+  private List<WifiP2pDevice> peerDevices = new ArrayList<WifiP2pDevice>();
+  private InetAddress selectedPeerDeviceInetAddress;
+  private InetAddress fileReceiverDeviceAddress;  // IP address of the file receiving device
+  private boolean fileTransferStarted = false;
+
+  private PeerGroupHandshakeAsyncTask peerGroupHandshakeAsyncTask;
+  private SenderDeviceAsyncTask[] senderDeviceAsyncTaskArray;
+  private ReceiverDeviceAsyncTask receiverDeviceAsyncTask;
+
+  @Override
+  protected void onCreate(@Nullable Bundle savedInstanceState) {
+    KiwixApplication.getApplicationComponent().activityComponent()
+        .activity(this)
+        .build()
+        .inject(this);
+    setTheme(sharedPreferenceUtil.nightMode() ? R.style.Theme_AppCompat_DayNight_NoActionBar : R.style.AppTheme);
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_local_file_transfer);
+    ButterKnife.bind(this);
+
+    /*
+     * Presence of file Uris decides whether the device with the activity open is a sender or receiver:
+     * - On the sender device, this activity is started from the app chooser post selection
+     * of files to share in the Library
+     * - On the receiver device, the activity is started directly from within the 'Get Content'
+     * activity, without any file Uris
+     * */
+    Intent filesIntent = getIntent();
+    fileUriArrayList = filesIntent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);
+    if (fileUriArrayList != null && fileUriArrayList.size() > 0) {
+      setDeviceAsFileSender();
+    }
+
+    setSupportActionBar(actionBar);
+    actionBar.setNavigationIcon(R.drawable.ic_close_white_24dp);
+    actionBar.setNavigationOnClickListener(new View.OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        wifiDirectManager.closeLocalFileTransferActivity();
```

Comment:
should call finish here instead of delegating to wifiDirectManager. It is the responsibility of the activity to finish itself

Reply:
The wifi direct connection has to be disconnected before closing the activity
Should I shift the method back to LocalFileTransferActivity and use the wifiDirectManager from there to call disconnect()?

Reply:
so once a person clicks on the `x` icon you should call either `finish()` or `onBackPressed()` then in `onDestroy` you forward the event on to WifiDirectManager, it probably makes sense to have some symmetry in method naming so if in WDM you call `start` to initialise it then you should call `stop` to clean up it's resources and these should be invoked in the appropriate lifecycle pair `onCreate/onDestroy, onPause/OnResume` you will see this repeated through android a lot like in the case of BroadcastReceivers

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/19b316644e88a12281b92b890d156bea07572ddd)
The previous code felt quite non-deterministic even to me, but now everything has gotten so streamlined

Reply:
It is always satisfying to see code coalesce into order